examples/generative/bayesian_neural_networks_wine.py

```diff
@@ -0,0 +1,423 @@
+"""
+
+"""
+## Introduction
+
+Taking a probabilistic approach to deep learning allows to account for *uncertainty*,
+so that models can assign less levels of confidence to incorrect predictions.
+Sources of uncertainty can be found in the data, due to measurement error or
+noise in the labels, or the model, due to insufficient data availability for
+the model to learn effectively.
+
+
+This example demonstrates how to build basic probabilistic Bayesian neural networks
+to account for these two types of uncertainty.
+We use [TensorFlow Probability](https://www.tensorflow.org/probability) library,
+which is compatible with Keras API.
+
+This example requires TensorFlow 2.3 or higher.
+You can install Tensorflow Probability using the following command:
+
+```python
+pip install tensorflow-probability
+```
+"""
+
+"""
+## The dataset
+
+We use the [Wine Quality](https://archive.ics.uci.edu/ml/datasets/wine+quality)
+dataset, which is available in the [TensorFlow Datasets](https://www.tensorflow.org/datasets/catalog/wine_quality).
+We use the red wine subset, which contains 4,898 examples.
+The dataset has 11numerical physicochemical features of the wine, and the task
+is to predict the wine quality, which is a score between 0 and 10.
+In this example, we treat this as a regression task.
+
+You can install TensorFlow Datasets using the following command:
+
+```python
+pip install tensorflow-datasets
+```
+"""
+
+"""
+## Setup
+"""
+
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+import tensorflow_datasets as tfds
+import tensorflow_probability as tfp
+
+"""
+## Create training and evaluation datasets
+
+Here, we load the `wine_quality` dataset using `tfds.load()`, and we convert
+the target feature to float. Then, we shuffle the dataset and split it into
+training and test sets. We take the first `train_size` examples as the train
+split, and the rest as the test split.
+"""
+
+
+def get_train_and_test_splits(train_size, batch_size=1):
+    dataset = (
+        tfds.load(name="wine_quality", as_supervised=True, split="train")
+        .shuffle(buffer_size=dataset_size)
+        .map(lambda x, y: (x, tf.cast(y, tf.float32)))
+    )
+
+    train_dataset = (
+        dataset.take(train_size)
+        .shuffle(buffer_size=train_size)
+        .batch(batch_size)
+        .repeat(1)
+    )
+    test_dataset = dataset.skip(train_size).batch(batch_size).repeat(1)
+
+    return train_dataset, test_dataset
+
+
+"""
+## Compile, train, and evaluate the model
+"""
+
+hidden_units = [8, 8]
+learning_rate = 0.001
+
+
+def run_experiment(model, loss, train_dataset, test_dataset):
+
+    model.compile(
+        optimizer=keras.optimizers.RMSprop(learning_rate=learning_rate),
+        loss=loss,
+        metrics=[keras.metrics.RootMeanSquaredError()],
+    )
+
+    print("Start training the model...")
+    model.fit(train_dataset, epochs=num_epochs)
```

Comment:
Need a validation dataset to monitor during training.

Reply:
Done